File: src/main/java/com/qtech/bubbles/common/entity/DamageSourceType.java
package com.qtech.bubbles.common.entity;

import java.util.Objects;

public class DamageSourceType {
    // Types
    public static final DamageSourceType COLLISION = new DamageSourceType("collision");
    public static final DamageSourceType POISON = new DamageSourceType("poison");
    public static final DamageSourceType UNKNOWN = new DamageSourceType("unknown");

    private final String name;

    protected DamageSourceType(String name) {
        this.name = name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DamageSourceType that = (DamageSourceType) o;
        return name.equals(that.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }
}

File: src/main/kotlin/com/qtech/bubbles/common/entity/DamageSourceType.kt
package com.qtech.bubbles.common.entity

import java.util.*

open class DamageSourceType protected constructor(private val name: String) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as DamageSourceType
        return name == that.name
    }

    override fun hashCode(): Int {
        return Objects.hash(name)
    }

    companion object {
        // Types
        @JvmField
        val COLLISION = DamageSourceType("collision")
        @JvmField
        val POISON = DamageSourceType("poison")
        val UNKNOWN = DamageSourceType("unknown")
        val GENERIC = DamageSourceType("generic")
    }
}